src/app/InteractionModelDelegate.h

```diff
@@ -0,0 +1,156 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the classes corresponding to CHIP Interaction Model Delegate.
+ *
+ */
+
+#pragma once
+
+#include <core/CHIPCore.h>
+#include <messaging/ExchangeContext.h>
+#include <system/SystemPacketBuffer.h>
+
+namespace chip {
+namespace app {
+
+struct EventPathParams;
+class ReadHandler;
+
+/**
+ * @brief
+ *   This class provides a skeleton for the callback functions. The functions will be
+ *   called by InteractionModel on specific events. If the user of InteractionModel
+ *   is interested in receiving these callbacks, they can specialize this class and handle
+ *   each trigger in their implementation of this class.
+ */
+class InteractionModelDelegate
+{
+public:
+    virtual ~InteractionModelDelegate() {}
+
+    enum class EventId
+    {
+        // InteractionModelEngine
+        kIncomingInvokeCommandRequest,
+        kIncomingReadRequest,
+        kIncomingReportRequest,
+        // ReadClient
+        kReadRequestPrepareNeeded,
+        kReportDataRequest,
+        kReportProcessed,
+        kEventStreamReceived,
+        // ReadHandler
+        kReadRequestParsed,
+    };
+
+    /**
+     * Incoming parameters sent with events generated directly from this component
+     *
+     */
+    union InEventParam
+    {
+        InEventParam() { Clear(); }
+
+        void Clear(void) { memset(this, 0, sizeof(*this)); }
+
+        struct
+        {
+            // ExchangeContext managed externally and should not be closed by event processing.
+            const PacketHeader * packetHeader;
+        } mIncomingInvokeCommandRequest;
```

Comment:
just remove it, would add in next pr after we merge interactionModelDelegate.h, current immediate requirement from zcl side is that we need to add InvokeCommandResponse callback so that zcl can know when the message has been processed successfully, then do others.